quintoandar_kafka/kafka_consumer.py

```diff
@@ -3,53 +3,55 @@
 
 from retrying import retry
 from kafka.consumer.group import KafkaConsumer
+from quintoandar_kafka import IdempotenceClient
+from functools import partial
 
 
-class KafkaConsumerClient:
-
-    def __init__(self, group_id, bootstrap_servers, topic,
-                 processor, deserializer=None, idempotence_client=None):
+class KafkaDefaultConsumer():
+    def __init__(self, processor, **kwargs):
         self.log = logging.getLogger(__name__)
-        self.topic = topic
         self.processor = processor
-        self.deserializer = deserializer or self.default_deserializer
-        self.idempotence_client = idempotence_client or\
-            DefaultIdempotenceClient()
-        self.consumer = self.connect(group_id, bootstrap_servers)
-        self.consumer.subscribe(topic)
+        if kwargs.get('deserializer') is None:
+            kwargs['deserializer'] = self.default_deserializer
+        if kwargs.get('auto_offset_reset') is None:
+            kwargs['auto_offset_reset'] = 'latest'
+        self.consumer = self.connect(kwargs)
+        self.consumer.subscribe(kwargs['topic'])
 
     @retry(stop_max_attempt_number=10, wait_fixed=3000)
-    def connect(self, group_id, bootstrap_servers):
-        return KafkaConsumer(
-            value_deserializer=self.deserializer,
-            group_id=group_id,
-            bootstrap_servers=bootstrap_servers,
-            auto_offset_reset='latest')
+    def connect(self, kwargs):
+        return KafkaConsumer(kwargs)
+
+    def default_deserializer(self, m):
+        try:
+            return json.loads(m.decode('utf8'))
+        except Exception as ex:
+            self.log.error("Failed to decode message: %s", ex, exc_info=True)
+            return {}
 
     def start(self):
         for message in self.consumer:
             if not message.value:
                 continue
 
-            if not self.idempotence_client.is_unique(self.topic, message):
-                continue
-
             self.processor(message)
 
-            self.idempotence_client.mark_consumed_message(self.topic, message)
 
-    def default_deserializer(self, m):
-        try:
-            return json.loads(m.decode('utf8'))
-        except Exception as ex:
-            self.log.error("Failed to decode message: %s", ex, exc_info=True)
-            return {}
+class KafkaIdempotentConsumer(KafkaDefaultConsumer):
 
+    def __init__(self, redis_host, redis_port, processor,
+                 key_extractor, **kwargs):
```

Comment:
I'm not sure I follow, do we need both a `processor` and a `key_extractor`? I had understood that only one custom func for extracting the messages and handling it would be necessary. Let me try to write some example/pseudo code:

```python
from quintokafka.consumers import KafkaIdempotentConsumer.

def Main():

    ...

        consumer = KafkaIdempotentConsumer(
            redis_host = '...',
            redis_port = '...',
            key_extractor = self.key_extractor
        )

    def key_extractor(message):
        event = message.value.get('eventName')

        if event in [ 'send_message', 'USER_PREFERENCES_UPDATE', 'USER_PREFERENCES_CREATE' ]:
            return message.value
        if event in [ 'check_contacts' ]:
            return message      
```

Wouldn't this work ? Then the "processor" code would be default for everyone.

Reply:
By the way: https://kafka-python.readthedocs.io/en/master/usage.html

There is a key_deserializer there as well... just wondering if it ties to our key_extractor somehow? :thinking: 

Maybe we should rename key_extractor to key_deserializer? Or if they are completely diferent things we should find a better name then key_extractor, because they are essentially the same "name". Maybe `idempotent_key`?

Maybe there's a default name for those things.

Reply:
They are not the some thing. I guess idempotent_key is fine.